Assets/LeapMotion/Core/Scripts/DataStructures/Maybe.cs

```diff
@@ -85,20 +85,27 @@ public struct Maybe<T> : IEquatable<Maybe<T>>, IComparable, IComparable<Maybe<T>
     private readonly T _t;
 
     /// <summary>
-    /// Constructs a Maybe given a value.  If the value is non-null, this maybe
-    /// will have a value.  If the value is null, this maybe will have no value.
+    /// Constructs a Maybe given a value. If the value is not null, this maybe will have
+    /// a value. If the value is null, this maybe will have no value. For value types,
+    /// the Maybe struct will always have a value. (Use Maybe.None to refer to "no value.")
     /// </summary>
     public Maybe(T t) {
-      hasValue = t != null;
+      if (Type<T>.isValueType) {
+        hasValue = true;
+      }
+      else {
+        hasValue = t != null;
+      }
+
       _t = t;
     }
 
     /// <summary>
-    /// Constructs a Maybe given a specific value.  This value needs to always be
-    /// non-null.
+    /// Constructs a Maybe given a specific value. This value needs to always be
+    /// non-null if the type is a reference type.
     /// </summary>
     public static Maybe<T> Some(T t) {
-      if (t == null) {
+      if (Type<T>.isValueType && t == null) {
```

Comment:
I think this needs to be `!Type<T>.isValueType`.  Currently the check is always false, since being a value type and being null are mutually exclusive.